### ISSUE TITLE
Struct passing on ARM 32-bit

### ISSUE THREAD

**Issue by manu-silicon** [dotnet/coreclr#3754]
I have the following C# code:

```
        public static ValueParameterKey<T> NewValue<T>(T defaultValue = default(T), string name = null) where T : struct
        {
            if (name == null)
                name = string.Empty;

            return new ValueParameterKey<T>(name, 1, new ParameterKeyValueMetadata<T>(defaultValue));
        }
```

which causes the following assert on line 4124 of compiler.cpp to fail:

```
Assert failure(PID 12411 [0x0000307b], Thread: 12411 [0x307b]): Assertion failed '(i > 0) || (regNum == varDsc->lvArgReg)' in 'ParameterKeys:NewValue(struct,ref):ref' (IL size 34)

    File: /home/ubuntu/local/Microsoft/coreclr/src/jit/codegencommon.cpp Line: 4124
    Image: /usr/local/home/ubuntu/work/SpaceEscape/corerun
```

When it fails ```regNum``` is ```REG_RO``` while ```varDsc->lvArgReg``` is 16. In varDsc, the following fields are set to 1:
```
lvIsHfaRegArg = 1 '\001'
lvHfaTypeIsFloat = 1 '\001'
```

Any hints on how to best fix this?

I've attached the output with COMPlus_JitDump=NewValue

[output.txt](https://github.com/dotnet/coreclr/files/175553/output.txt)

**Comment by jkotas** [comment]
cc @dotnet/jit-contrib @myungjoo @leemgs

**Comment by BruceForstall** [comment]
There have been a lot of changes in this area due to Linux x64 and ARM64 calling conventions, with work under FEATURE_UNIX_AMD64_STRUCT_PASSING and FEATURE_MULTIREG_ARGS. It's probable one of those many changes regressed ARM32. You'll have to do some more detailed debugging here. If you provide a self-contained repro, maybe someone can look at it. I can't determine the issue from what I see here.

**Comment by manu-silicon** [comment]
@BruceForstall I've attached the C# code that reproduces the issue.

[Program.txt](https://github.com/dotnet/coreclr/files/177122/Program.txt)

**Comment by manu-silicon** [comment]
For reference, the initial output.txt had a different failure which was due to issue #3778 which was just fixed.

**Comment by leemgs** [comment]
> https://github.com/dotnet/coreclr/files/177122/Program.txt

@manu-silicon , BTW,  Why did you assign `var p = NewValue(q)` statement? It is not never used in C# code. What for?

**Comment by manu-silicon** [comment]
@leemgs Old habit to never hide a result. I did not try with a procedure to see if it would fail too.

**Comment by leemgs** [comment]
> https://github.com/dotnet/coreclr/files/177122/Program.txt

@manu-silicon , I tried to run your C# code to analyze the issue on my Linux/ARM emulator. It's strange. The result is "No reponse" as following: 

```
emul@arm32#> ./corerun -c ./ ../../example/struct-passing.exe

           . . . . Waiting for . . . .  ( <--- This is just my comment. )
^C
real	17m29.228s
user	10m25.470s
sys	6m59.480s
emul@arm32#>
emul@arm32#>

```

**Comment by manu-silicon** [comment]
Are you in debug mode? If not, it will do an infinite loop.

**Comment by leemgs** [comment]
@manu-silicon , My test environment is as following: Target is Linux/ARM Emulaotor (ARMv7le-hard-fp),  CoreCLR was compiled by debug mode.

**Comment by manu-silicon** [comment]
I'm testing directly on a Linux/ARM board (NVidia Tegra TK1).

**Comment by manu-silicon** [comment]
Making some progress here. I believe I found where the regression took place. I'm going to test a fix and see if it works.

**Comment by manu-silicon** [comment]
See PR #3803.

**Comment by manu-silicon** [comment]
Closing as merged in #3803
